libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinPlugin.kt

```diff
@@ -741,6 +741,7 @@ abstract class AbstractAndroidProjectHandler(private val kotlinConfigurationTool
             ).apply {
                 kotlin.srcDir(project.file(project.file("src/${sourceSet.name}/kotlin")))
                 kotlin.srcDirs(sourceSet.java.srcDirs)
+                sourceSet.java.srcDir(project.file("src/$name/kotlin"))
```

Comment:
Having this path declared here might cause inconsistency if the path changes in the other places where it's used.
Kotlin source sets have their `srcDirs` initialized here: https://github.com/JetBrains/kotlin/blob/0c9d962f9f474479fd6328df9e13e539e25a31b5/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/sources/KotlinSourceSetFactory.kt#L36
I think it would be safer not to duplicate the path pattern here and reuse the existing content of the `srcDirs`. For example, this could be done as follows:
```
sourceSet.java.srcDirs(*kotlin.srcDirs.minus(sourceSet.java.srcDirs).toTypedArray())
```
(the `.minus(...)` part is a safety measure to avoid duplicating the Java source directories, for instance, if some plugin or a build script uses both collections to pack a sources JAR)

Reply:
Oh, this is pretty clever, didn't think about that! The src folder was just a given in my head! 
I am just a little bit afraid, that this alone (without solving the test source folder conflict) might introduce some strange behavior. With only that line we will add 
- "src/androidTest/kotlin" to Android's "test" sourceSet
- "src/androidTest/kotlin" to Android's "androidTest" sourceSet

However: If we move that line before adding the kotlin source folder" 
`kotlin.srcDir(project.file(project.file("src/${sourceSet.name}/kotlin")))`
Then I expect it to be correct!

Reply:
Pushed this change in the commit that also handles the conflict as you proposed. I now think it only makes sense to ship this with conflict resolution.